proposals/Fleetctl-trigger.md

```diff
@@ -0,0 +1,95 @@
+# Fleetctl trigger
+
+## Goal
+
+As a user, I would like to trigger a set of async jobs using `fleetctl`. For example, I’d like to
+trigger a vuln scan, or an MDM dep sync. 
+
+The proposed solution to accomplish this goal enables a new CLI command: 
+`fleetctl trigger --name <NAME>`.   
+
+## Background
+
+Currently, the Fleet server uses the `schedule` package to create sets of defined jobs that are run
+serially at defined intervals. The initial schedule interval must be specified at the point the
+schedule is instantiated via `schedule.New`. Optionally, `schedule. WithConfigReloadInterval`
+accepts a reload interval function. If specified, the reload interval function is periodically
+called and its return value becomes the new the schedule interval. This mechanism allows the
+schedule interval to be modified by user, for example, by changing the app config; however, there no
+mechanism currently to trigger async jobs on an ad hoc basis.
+
+## Proposal
+
+### New CLI command `fleetctl trigger --name <NAME>`
+- Upon this command, the CLI client makes a request to a new authenticated endpoint (see below) to
+  trigger an ad hoc run of the named schedule. 
+
+### New `schedule` option `WithTrigger` 
```

Comment:
Do we need this option? I would think all schedules could be triggered ie it should just be created as part of `schedule.New`. Also, if we have a `schedules` map on the `Service` struct, then the schedule can just expose the ability to trigger it eg add a new method `Trigger()` to `Schedule`.

Reply:
Yeah, I thought that was another way we could go with this. 

I'm curious if anyone can think of any use cases where we would not want a particular schedule to be triggerable?

Reply:
Cannot think of one. +1 to Michal's idea.

Reply:
@lucas, What about statistics and counts that get tracked over time, for example? 

Would those need to be protected so that triggered runs don't skew trends? Do we leave it to the dev implementing each job to remember to protect this independently? For example, I believe the usage stats rely on the `updated_at` timestamp from the datastore. Is there additional benefit to defaulting to the no-trigger behavior to make it more likely that potential side-effects are taken into account when new features are being implemented?

Reply:
> What about statistics and counts that get tracked over time, for example?

Do we have this implemented? Totally missed that :)

Reply:
> Do we have this implemented? Totally missed that :)

@lucasmrod currently tracking over time is mostly happening on the fleetdm.com side with the usage stats. The Fleet server is sending stats on an at-most-once per 24-hours basis and we're enforcing that via the `updated_at` timestamp in the `statistics` table. It seems like there is movement toward exposing trend data that involves more time-based derivation by the Fleet server, which I can see being incorporated into the UI. For example, tracking [policy violation-days](https://github.com/fleetdm/fleet/issues/6027). It won't be hard to manage these using `updated_at` timestamps either.

Reply:
But my bigger concern is how our design choices for the scheduler API now influence what a future developer assumes  about the how the schedule interval behaves and what sort of guarantee the interval provides by default (e.g., at-most-once-per-interval, at-least-once-per-interval, only-once-per-interval).

In general, I think the concept of a schedule implies that scheduled things happen only once per interval. 

By having the `WithTrigger` option, we're making it more explicit that this particular instantiation of the schedule abstraction is departing from that natural assumption about behavior.

In my opinion, the default behavior of the API ought to reflect the natural assumptions implied by the abstraction.

Reply:
Hard to know right now what's best...

I'm leaning more on the side of:
At the package level, we could define `Schedule`s in our `schedule` package to be "trigger-able" by default (via a `Trigger()` method or something) and have an option to disable such feature (in case that in the future we want a schedule to not be trigger-able).

(Properly documenting the above in the schedule package)